### PR TITLE
Update docker_image_build.py

### PR DESCRIPTION
**Pull request by bitsky6**
##### SUMMARY

Added new options
- push
- pull
- secret
- load
- multi-platform

Secret accepts multiple from password and files
multi-platform now accepts as a list of platforms

##### ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- New Module Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/modules/docker_image_build.py [hunk 1]
```diff
@@ -14,10 +14,11 @@
 
 short_description: Build Docker images using Docker buildx
 
-version_added: 3.6.0
+version_added: "3.8.0"
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:17]: Please undo the change. The module was added to version 3.6.0 of the collection.

File: plugins/modules/docker_image_build.py [hunk 2]
```diff
@@ -14,10 +14,11 @@
...
 
 description:
-  - This module allows you to build Docker images using Docker's buildx plugin (BuildKit).
+  - This module allows you to build Docker images using Docker's buildx plugin (BuildKit),
+    supporting features like multi-platform builds, secrets, and conditional image loading or pushing.
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:21]:
```suggestion
  - This module allows you to build Docker images using Docker's buildx plugin (BuildKit).
    It supports features such as multi-platform builds, secrets, and conditional image loading or pushing.
```

File: plugins/modules/docker_image_build.py [hunk 3]
```diff
@@ -89,7 +90,10 @@
     type: str
   platform:
     description:
-      - Platform in the format C(os[/arch[/variant]]).
+      - Target platform(s) for the build, specified as a single string or a list of strings.
+      - Each platform string should be in the format "os/arch[/variant]".
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:94]:
```suggestion
      - Each platform string should be in the format C(os/arch[/variant]).
```

File: plugins/modules/docker_image_build.py [hunk 4]
```diff
@@ -89,7 +90,10 @@
...
+      - Example single platform "linux/amd64".
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:95]:
```suggestion
      - For example, a single platform is C(linux/amd64).
```

File: plugins/modules/docker_image_build.py [hunk 5]
```diff
@@ -89,7 +90,10 @@
...
+      - Example multiple platforms ["linux/amd64", "linux/arm64/v8"].
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:96]: This example is invalid, this option does not accept multiple values right now.

File: plugins/modules/docker_image_build.py [hunk 6]
```diff
@@ -110,6 +114,38 @@
       - never
       - always
     default: never
+  secret:
+    description: Secrets to expose to the build.
+    type: list
+    elements: dict
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:120]:
```suggestion
    elements: dict
    version_added: 3.9.0
```

File: plugins/modules/docker_image_build.py [hunk 7]
```diff
@@ -110,6 +114,38 @@
...
+    suboptions:
+      id:
+        description: The secret identifier.
+        type: str
+        required: True
+      src:
+        description: Source path of the secret.
+        type: str
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:128]:
```suggestion
        type: path
```

File: plugins/modules/docker_image_build.py [hunk 8]
```diff
@@ -110,6 +114,38 @@
...
+      value:
+        description: Value of the secret.
+        type: str
+      type:
+        description: Type of the secret.
+        type: str
+        choices: ['file', 'password']
+        required: True
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:136]:
```suggestion
        required: true
```

File: plugins/modules/docker_image_build.py [hunk 9]
```diff
@@ -110,6 +114,38 @@
...
+  load:
+    description:
+      - Load the built image into Docker's local image store.
+      - Cannot be used together with C(push).
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:140]:
```suggestion
      - Cannot be used together with O(push).
```

File: plugins/modules/docker_image_build.py [hunk 10]
```diff
@@ -110,6 +114,38 @@
...
+    type: bool
+    default: true
+  push:
+    description:
+      - Push the built image to a Docker registry.
+      - Cannot be used together with C(load).
+    type: bool
+    default: false
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:148]:
```suggestion
    default: false
    version_added: 3.9.0
```

File: plugins/modules/docker_image_build.py [hunk 11]
```diff
@@ -110,6 +114,38 @@
...
+      src:
+        description: Source path of the secret.
+        type: str
+      value:
+        description: Value of the secret.
+        type: str
+      type:
+        description: Type of the secret.
+        type: str
+        choices: ['file', 'password']
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:135]:
```suggestion
        choices:
          - file
          - password
```

File: plugins/modules/docker_image_build.py [hunk 12]
```diff
@@ -110,6 +114,38 @@
...
+  secret:
+    description: Secrets to expose to the build.
+    type: list
+    elements: dict
+    suboptions:
+      id:
+        description: The secret identifier.
+        type: str
+        required: True
+      src:
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:126]: Why not call this argument `path`?

File: plugins/modules/docker_image_build.py [hunk 13]
```diff
@@ -128,6 +164,30 @@
...
     dockerfile: Dockerfile-3.12
+
+- name: Build a multi-platform image
+  your_module_name:
+    name: my-multi-platform-image
+    path: /path/to/context
+    platform:
+      - linux/amd64
+      - linux/arm64/v8
+    load: true
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:175]:
```suggestion
```
I would not specify default values in examples.

File: plugins/modules/docker_image_build.py [hunk 14]
```diff
@@ -128,6 +164,30 @@
...
     path: /home/user/images/python
     dockerfile: Dockerfile-3.12
+
+- name: Build a multi-platform image
+  your_module_name:
+    name: my-multi-platform-image
+    path: /path/to/context
+    platform:
+      - linux/amd64
+      - linux/arm64/v8
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:174]: `platform` is a string parameter, not a list parameter.

File: plugins/modules/docker_image_build.py [hunk 15]
```diff
@@ -248,35 +323,86 @@ def add_args(self, args):
...
+                args.extend(['--secret', secret_str])
+        if self.load:
+            args.append('--load')
+        if self.push:
+            args.append('--push')
+
+    def parse_platforms(self, platform_param):
+        platforms = []
+        if platform_param:
+            if isinstance(platform_param, list):
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:344]: This can never be true, the value is always `None` or a string.

File: plugins/modules/docker_image_build.py [hunk 16]
```diff
@@ -248,35 +323,86 @@ def add_args(self, args):
...
+                platforms = platform_param
+            elif isinstance(platform_param, str):
+                try:
+                    parsed = json.loads(platform_param)
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:348]: Do not JSON parse strings that shouldn't be JSON.

Change the parameter type to `list` with `elements=str`. Then it's always a list of string or `None`.

File: plugins/modules/docker_image_build.py [hunk 17]
```diff
@@ -248,35 +323,86 @@ def add_args(self, args):
...
+                with os.fdopen(temp_fd, 'w') as tmp:
+                    tmp.write(str(secret['value']))
+                secret['src'] = temp_path
+                temp_files.append(temp_path)
+
+            elif 'src' in secret and secret['type'] == 'file':
+                if not os.path.isfile(secret['src']):
+                    self.fail("Secret file {0} not found.".format(secret['src']))
+            else:
+                self.fail("Secret must include either a 'value' or a 'src' key.")
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:374]: This should be handled by argument parsing, not by code.

File: plugins/modules/docker_image_build.py [hunk 18]
```diff
@@ -248,35 +323,86 @@ def add_args(self, args):
...
+                    platforms.append(platform_param)
+            else:
+                self.fail("Invalid platform format. Expected string, JSON list, or YAML list.")
+        return platforms
+
+    def handle_secrets(self):
+        secrets = self.secret or []
+        temp_files = []
+        for secret in secrets:
+            if 'value' in secret:
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:363]: This is **always** true. `secret['value']` is either a string or `None`, depending on whether the user provided a value here or not.

File: plugins/modules/docker_image_build.py [hunk 19]
```diff
@@ -248,35 +323,86 @@ def add_args(self, args):
...
+                temp_fd, temp_path = tempfile.mkstemp()
+                with os.fdopen(temp_fd, 'w') as tmp:
+                    tmp.write(str(secret['value']))
+                secret['src'] = temp_path
+                temp_files.append(temp_path)
+
+            elif 'src' in secret and secret['type'] == 'file':
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:370]: Why is `secret['type']` required here, but not above?

File: plugins/modules/docker_image_build.py [hunk 20]
```diff
@@ -248,35 +323,86 @@ def add_args(self, args):
...
+                    self.fail('Building %s:%s failed' % (self.name, self.tag), stdout=to_native(stdout), stderr=to_native(stderr))
+
+                results.update(stdout=to_native(stdout), stderr=to_native(stderr), image=self.client.find_image(self.name, self.tag) or {})
+
+        finally:
+            for temp_path in temp_files:
+                try:
+                    os.remove(temp_path)
+                except OSError as e:
+                    self.fail(msg='Error cleaning up temporary file {0}: {1}'.format(temp_path, str(e)))
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:405]: Cleanup should be handled by `AnsibleModule`, not by custom code. Check out `AnsibleModule.add_cleanup_file()`.

File: plugins/modules/docker_image_build.py [hunk 21]
```diff
@@ -248,35 +323,86 @@ def add_args(self, args):
...
+            if not self.check_mode:
+                args = ['buildx', 'build', '--progress', 'plain']
+                self.add_args(args)
+                args.extend(['--', self.path])
+                rc, stdout, stderr = self.client.call_cli(*args)
+
+                if rc != 0:
+                    self.fail('Building %s:%s failed' % (self.name, self.tag), stdout=to_native(stdout), stderr=to_native(stderr))
+
+                results.update(stdout=to_native(stdout), stderr=to_native(stderr), image=self.client.find_image(self.name, self.tag) or {})
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:398]: Please keep closer to the original code instead of rewriting everything. Thanks.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Please note that you have to add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment).

File: plugins/modules/docker_image_build.py [hunk 22]
```diff
@@ -110,6 +114,38 @@
...
+        description: Type of the secret.
+        type: str
+        choices: ['file', 'password']
+        required: True
+  load:
+    description:
+      - Load the built image into Docker's local image store.
+      - Cannot be used together with C(push).
+    type: bool
+    default: true
```
Comment (felixfontein) [inline plugins/modules/docker_image_build.py:142]: Why is this default `true`? Isn't this a breaking change?

Comment (bitsky6) [reply plugins/modules/docker_image_build.py:143]: I usually prefer by default store it local but will remove default so who want can specifiy

Comment (bitsky6) [reply plugins/modules/docker_image_build.py]: changed it to false

**Comment by felixfontein** [comment]
@bitsky6 there are several unresolved review comments. Can you please take a look at these and address them? Thanks.

**Comment by felixfontein** [comment]
I created another PR that adds similar features in #852.
